File: 1. Path Sum II.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def pathSum(self, root, targetSum):
        # Time Complexity - O(n)
        # Space Complexity - O(h) - as the max number of elements in the stack at any time is the height of the tree

        self.result = []
        if not root:
            return self.result
        self.helper(root, targetSum, 0, [])
        return self.result

    def helper(self, root, targetSum, sum_, path):
        # Base Condition
        if root is None:
            return

        sum_ += root.val
        path.append(root.val)

        if root.left is None and root.right is None:
            if sum_ == targetSum:
                self.result.append(path.copy())

        self.helper(root.left, targetSum, sum_, path)
        self.helper(root.right, targetSum, sum_, path)
        path.pop()

File: 1pathSumII.java
// this approach has a problem because arraylist is pass by reference so it doesn't 'preserve' the state when one recursive call ends
// we can solve this in multiple ways

class Solution {
    List<List<Integer>> result;
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        result = new ArrayList<>();
        if(root == null) {
            return result;
        }
        helper(root, 0, targetSum, new ArrayList<>());
        return result;
    }
    private void helper(TreeNode root, int curSum, int targetSum, List<Integer> path) {
        if(root == null) {
            return;
        }
        curSum = curSum + root.val;
        path.add(root.val);
        if(root.left == null && root.right == null) {
            if(curSum == targetSum) {
                result.add(path);
            }
        }
        helper(root.left, curSum, targetSum, path);
        helper(root.right, curSum, targetSum, path);
    }
}